spark/src/main/scala/org/apache/hadoop/fs/versioned/VersionedFileSystem.scala

```diff
@@ -0,0 +1,138 @@
+package org.apache.hadoop.fs.versioned
+
+import java.net.URI
+import java.util.Objects
+
+import com.gu.tableversions.metastore.VersionPaths
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs._
+import org.apache.hadoop.fs.permission.FsPermission
+import org.apache.hadoop.util.Progressable
+
+class VersionedFileSystem extends FileSystem {
+
+  private var baseURI: URI = _
+  private var baseFS: FileSystem = _
+  private var mappings: Map[String, String] = _
+  private var version: String = _
```

Comment:
I think we'd want to make these thread safe.

Reply:
Don't think it's needed. AFAICT, spark never reruns `initialize` on a previously existing object, and our own methods do not modify them

Reply:
Strictly speak, a value that's assigned to these `var`s in `initialize` may not be visible to other threads reading it later. It may be unlikely to be a problem, but I'd rather be 100% sure. Just adding `@volatile` annotations should be enough to fix it.

Reply:
I don't really like coding defensively as there might be nothing to fix here.

I checked a few of the other spark FS and they all have plenty of private vars being assigned upon the `initialize` call. If spark was exhibiting race conditions due to the lack of explicit memory barriers, it most likely would show in the coding style of those other FS too.

I was also reading things like https://stackoverflow.com/questions/1850270/memory-effects-of-synchronization-in-java which dwelves arond a similar code in `FileSystem` where `fs` is created out of `synchronized` blocks but should be flushed alongside the `CACHE` variable as the later refers to the former.

Reply:
Not knowing too much about Java concurrency, is there a downside to adding `@volatile`? It might not strictly be necessary, but it at least guarantees (as far as I can tell) that these vars would be available to all threads after assignment. It's true that other `FileSystem` implementations mutate variables in `initialise` without the annotation, but if it doesn't have an overhead I'd be tempted to add it defensively.

Reply:
The only downside I can see is a tiny overhead in the added memory barrier, but I would imagine that's on many orders of magnitude smaller than the actual I/O operations these methods represent. It probably doesn't make a difference in practice but I would tend to prefer correctness over performance by default.

Reply:
We should refrain from coding defensively. And favor consistency.

> [...] the rest of the code base doesn't use it so I'd rather change it.
> I tend to avoid a having too strong of a personal preference in favour of consistency with the rest of the codebase...

in this very same PR.

Reply:
Consistency with what? Are there other cases where we mutate `var`s that are accessible across threads?

Reply:
WIth other spark's filesystem implementations. That seems relevant to me.

Reply:
Hadoop code has a different set of practices, same for Spark code. As mentioned above, I prefer going with what's correct rather than what probably works and that other bits of code does. (Otherwise we'd be writing this filesystem implementation in Java for a start...)

I would also say that I don't think it's worth spending much time on debating this point.

Reply:
Indeed.